src/cascade/input_data/db/bundle.py

```diff
@@ -0,0 +1,327 @@
+        bundle_data = _get_bundle_data(execution_context, bundle_id, tier=2)
+        covariate_data = _get_study_covariates(execution_context, bundle_id, tier=2)
+        with cursor(execution_context) as c:
+            _upload_bundle_data_to_tier_3(c, model_version_id, bundle_data)
+            _upload_study_covariates_to_tier_3(c, model_version_id, covariate_data)
+        return True
+
+
+def _normalize_measures(data):
+    """Transform measure_ids into canonical measure names
+    """
+    data = data.copy()
+    data["measure"] = data.measure_id.apply(MEASURES.get)
+    return data
+
+
+def _normalize_sex(data):
+    """Transform sex_ids into strings
+    """
+    data = data.copy()
+    data["sex"] = data.sex_id.apply({1: "Male", 2: "Female", 3: "Both"}.get)
+    return data
+
+
+def _normalize_bundle_data(data):
+    """Normalize bundle columns, strip extra columns and index on `seq`
+    """
+    data = _normalize_measures(data)
+    data = _normalize_sex(data)
+
+    data = data.set_index("seq")
+
+    cols = ["measure", "mean", "sex", "standard_error", "age_start", "age_end", "year_start", "year_end"]
+
+    return data[cols]
+
+
+def _covariate_ids_to_names(execution_context, study_covariates):
+    """Convert study_covariate_ids to canonical study covariate names
+    """
+    query = """
+    select study_covariate_id, study_covariate
+    from epi.study_covariate
+    where study_covariate_id in %(covariate_ids)s
+    """
+    with cursor(execution_context) as c:
+        c.execute(query, args={"covariate_ids": tuple(study_covariates.study_covariate_id.unique())})
+        covariate_mapping = dict(list(c))
+
+    study_covariates = study_covariates.copy()
+    study_covariates["name"] = study_covariates.study_covariate_id.apply(covariate_mapping.get)
+
+    del study_covariates["study_covariate_id"]
+
+    return study_covariates
+
+
+def _normalize_covariate_data(execution_context, study_covariates):
+    study_covariates = _covariate_ids_to_names(execution_context, study_covariates)
+
+    study_covariates = study_covariates.set_index("seq")
+
+    return study_covariates.name
+
+
+def bundle_with_study_covariates(execution_context, bundle_id=None, tier=3):
+    """Get bundle data with associated study covariate labels.
+
+    Args:
+        execution_context (ExecutionContext): The context within which to make the query
+        bundle_id (int): Bundle to load. Defaults to the bundle associated with the context
+        tier (int): Tier to load data from. Defaults to 3 (frozen data) but will also accept 2 (scratch space)
+
+    Returns:
+        A tuple of (bundle data, study covariate labels) where the bundle data is a pd.DataFrame and the labels are a
+        pd.Series with an index aligned with the bundle data
+    """
+    if bundle_id is None:
+        bundle_id = _get_bundle_id(execution_context)
+
+    bundle = _get_bundle_data(execution_context, bundle_id, tier=tier)
+    bundle = _normalize_bundle_data(bundle)
```

Comment:
It has been mentioned to me in a PR that I shouldn't re-use a variable name, pointing to Code Complete for the justification.  

Any thoughts on that?

Reply:
It's something you want to be conscious of but I think there are times where it's useful. For example, this case where I'm transforming a value several times and I want to make the steps clear (ie. not have a huge, complicated composed function) but I don't need or want to keep references to the intermediate versions of the data.

I don't know why Code Complete would make a hard rule about this, that seems dumb and you can tell whoever the mysterious person that gave you that feed back that I said so. Not doing it has several advantages: I have fewer names to keep track of. I don't have to think of new names for the intermediate values which would never be used anyway. The garbage collector may now be free to reap the old version of the value which it could not do if I'd changed names.

If I were reusing the same variable name for completely unrelated values then I agree, that seems like bad practice.